plugins/amoeba/platforms/reference/src/SimTKReference/AmoebaReferenceMultipoleForce.cpp

```diff
@@ -907,6 +1181,396 @@ RealOpenMM AmoebaReferenceMultipoleForce::calculateElectrostaticPairIxn(const Mu
+            valJD += inducedDipoleRotationMatrix[ii][jj] * _inducedDipole[kIndex][jj];
+        }
+        qiUindI[ii] = valID;
+        qiUinpI[ii] = valIP;
+        qiUindJ[ii] = valJD;
+        qiUinpJ[ii] = valJP;
+    }
+
+    // The Qtilde intermediates (QI frame multipoles) for atoms I and J
+    RealOpenMM qiQI[9], qiQJ[9];
+    // Rotate the permanent multipoles to the QI frame.
+    qiQI[0] = particleI.charge;
+    qiQJ[0] = particleK.charge;
+    for (int ii = 0; ii < 3; ii++) {
+        RealOpenMM valI = 0.0;
+        RealOpenMM valJ = 0.0;
+        for (int jj = 0; jj < 3; jj++) {
+            valI += qiRotationMatrix1[ii][jj] * particleI.sphericalDipole[jj];
+            valJ += qiRotationMatrix1[ii][jj] * particleK.sphericalDipole[jj];
+        }
+        qiQI[ii+1] = valI;
+        qiQJ[ii+1] = valJ;
+    }
+    for (int ii = 0; ii < 5; ii++) {
+        RealOpenMM valI = 0.0;
+        RealOpenMM valJ = 0.0;
+        for (int jj = 0; jj < 5; jj++) {
+            valI += qiRotationMatrix2[ii][jj] * particleI.sphericalQuadrupole[jj];
+            valJ += qiRotationMatrix2[ii][jj] * particleK.sphericalQuadrupole[jj];
+        }
+        qiQI[ii+4] = valI;
+        qiQJ[ii+4] = valJ;
+    }
+
+    // The Qtilde{x,y,z} torque intermediates for atoms I and J, which are used to obtain the torques on the permanent moments.
+    RealOpenMM qiQIX[9] = {0.0, qiQI[3], 0.0, -qiQI[1], sqrtThree*qiQI[6], qiQI[8], -sqrtThree*qiQI[4] - qiQI[7], qiQI[6], -qiQI[5]};
+    RealOpenMM qiQIY[9] = {0.0, -qiQI[2], qiQI[1], 0.0, -sqrtThree*qiQI[5], sqrtThree*qiQI[4] - qiQI[7], -qiQI[8], qiQI[5], qiQI[6]};
+    RealOpenMM qiQIZ[9] = {0.0, 0.0, -qiQI[3], qiQI[2], 0.0, -qiQI[6], qiQI[5], -2.0*qiQI[8], 2.0*qiQI[7]};
+    RealOpenMM qiQJX[9] = {0.0, qiQJ[3], 0.0, -qiQJ[1], sqrtThree*qiQJ[6], qiQJ[8], -sqrtThree*qiQJ[4] - qiQJ[7], qiQJ[6], -qiQJ[5]};
+    RealOpenMM qiQJY[9] = {0.0, -qiQJ[2], qiQJ[1], 0.0, -sqrtThree*qiQJ[5], sqrtThree*qiQJ[4] - qiQJ[7], -qiQJ[8], qiQJ[5], qiQJ[6]};
+    RealOpenMM qiQJZ[9] = {0.0, 0.0, -qiQJ[3], qiQJ[2], 0.0, -qiQJ[6], qiQJ[5], -2.0*qiQJ[8], 2.0*qiQJ[7]};
+
+    // The field derivatives at I due to permanent and induced moments on J, and vice-versa.
+    // Also, their derivatives w.r.t. R, which are needed for force calculations
+    RealOpenMM Vij[9], Vji[9], VjiR[9], VijR[9];
+    // The field derivatives at I due to only permanent moments on J, and vice-versa.
+    RealOpenMM Vijp[3], Vijd[3], Vjip[3], Vjid[3];
+    RealOpenMM rInvVec[7];
+
+    RealOpenMM prefac = (_electric/_dielectric);
+    RealOpenMM rInv = 1.0 / r;
+
+    // The rInvVec array is defined such that the ith element is R^-i, with the
+    // dieleectric constant folded in, to avoid conversions later.
+    rInvVec[1] = prefac * rInv;
+    for(int i = 2; i < 7; ++i)
+        rInvVec[i] = rInvVec[i-1] * rInv;
+
+    RealOpenMM mScale = scalingFactors[M_SCALE];
+    RealOpenMM pScale = scalingFactors[D_SCALE];
```

Comment:
The code in the repo is working and passes all tests. The only weird part is that the dScale variable is set to PSCALE, and vice versa. To make the code more readable you may want to redefine dScale and pScale to have the proper meanings, but change all instances of pScale usage to dSclae, and dScale to pScale. This is trivial to do in context-aware editors like the excellent QT Creator. I will gladly do it tomorrow if you like. I've been traveling for over 24 hours now, so I don't trust myself to do it right now!



> On Aug 20, 2015, at 5:11 PM, peastman <notifications@github.com> wrote:
> 
> In plugins/amoeba/platforms/reference/src/SimTKReference/AmoebaReferenceMultipoleForce.cpp:
> 
> > +    RealOpenMM Vij[9], Vji[9], VjiR[9], VijR[9];
> > +    // The field derivatives at I due to only permanent moments on J, and vice-versa.
> > +    RealOpenMM Vijp[3], Vijd[3], Vjip[3], Vjid[3];
> > +    RealOpenMM rInvVec[7];
> > +
> > +    RealOpenMM prefac = (_electric/_dielectric);
> > +    RealOpenMM rInv = 1.0 / r;
> > +
> > +    // The rInvVec array is defined such that the ith element is R^-i, with the
> > +    // dieleectric constant folded in, to avoid conversions later.
> > +    rInvVec[1] = prefac * rInv;
> > +    for(int i = 2; i < 7; ++i)
> > +        rInvVec[i] = rInvVec[i-1] * rInv;
> > +
> > +    RealOpenMM mScale = scalingFactors[M_SCALE];
> > +    RealOpenMM pScale = scalingFactors[D_SCALE];
> Could you clarify what you mean? I tried swapping the definitions of eCoef and dCoef, but that didn't fix it. What change do I need to make?
> 
> —
> Reply to this email directly or view it on GitHub.
>